utils/block_proto/blocks.js

```diff
@@ -71,35 +120,167 @@ export function render(block, config) {
   const draw = svgjs.SVG(); // eslint-disable-line
 
   if (block in BLOCK_DEFS) {
-    BLOCK_DEFS[block].renderer(draw, BLOCK_DEFS[block], config);
+    const blockDef = BLOCK_DEFS[block];
+    if (blockDef.validValues && !blockDef.validValues.includes(config)) {
+      config = null;
+    }
+    const blockSize = blockDef.renderer(draw, BLOCK_DEFS[block], config, null);
+    draw.size(blockSize.width + 2 * GLOBAL_DEFS.margin,
+        blockSize.height + 2 * GLOBAL_DEFS.margin);
   }
 
   return draw.svg();
 }
 
 /**
- * Renders a number value block.
- * @param {Object} draw The svgjs draw object.
- * @param {Object} defs The definition of the main block kind.
- * @param {string?} config The config string.
+ * Renders a center-aligned text on top of a block.
+ * @param {!Object} draw The svgjs draw object.
+ * @param {string} text The text to be drawn.
+ * @param {number} shapeWidth The shape width.
+ * @param {number} shapeHeight The shape height.
+ * @param {string} color The text color.
+ * @param {?svgjs.Point} offset The offset of the block.
  */
-function renderNumber(draw, defs, config) {
-  const value = '3.14';
-  const textLength = value.length;
-  const rectWidth = 2 * GLOBAL_DEFS.padding +
+function renderCenterText(draw, text, shapeWidth, shapeHeight, color, offset) {
+  const textAnchorX = shapeWidth / 2;
+  const textAnchorY = shapeHeight - GLOBAL_DEFS.padding -
+      GLOBAL_DEFS.fontBaselineHeight;
+  draw.text(text)
+      .font(GLOBAL_DEFS.font)
+      .fill(color)
+      .move(offset.x + textAnchorX, offset.y + textAnchorY);
+}
+
+/**
+ * Renders a number or string value in a round rectangle shape.
+ * @param {!Object} draw The svgjs draw object.
+ * @param {!Object} blockDef The definition of the block kind.
+ * @param {?string} config The config string.
+ * @param {?svgjs.Point} offset The offset of the block. If it is null, the
+ *     block is the main block and will be positioned to the center of the SVG
+ *     canvas.
+ * @return {{width: number, height: number}} The calculated size of the block.
+ *     Margins are not included.
+ */
+function renderRoundRectValue(draw, blockDef, config, offset) {
+  const valueString = config || blockDef.defaultValue;
+  let textLength = valueString.length;
+  if (blockDef.delimiter) {
+    textLength += 2;
+  }
+  const shapeWidth = 2 * GLOBAL_DEFS.padding +
+      textLength * GLOBAL_DEFS.fontCharWidth;
+  const shapeHeight = 2 * GLOBAL_DEFS.padding + GLOBAL_DEFS.fontCharHeight;
+  const shapeOffset = offset ||
+      new svgjs.Point(GLOBAL_DEFS.margin, GLOBAL_DEFS.margin);
+
+  draw.rect(shapeWidth, shapeHeight)
+      .fill(blockDef.background)
+      .radius(GLOBAL_DEFS.rectRadius)
+      .move(shapeOffset.x, shapeOffset.y);
+
+  renderCenterText(draw, valueString, shapeWidth, shapeHeight, blockDef.color,
+      shapeOffset);
+
+  if (blockDef.delimiter) {
```

Comment:
Is it possible to add delimiters at the beginning and end of the valueString, then render the whole string with delimiters, to avoid the calculation of the delimiters' position?

Reply:
Delimiters have a separate text color different from the main text.